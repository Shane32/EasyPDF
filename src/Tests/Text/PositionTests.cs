using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Text;

public class PositionTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public PositionTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Theory]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.18f, 0.03f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1f, 6f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.18f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 6f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.216f, 0.036f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 0f, 1.05f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 1f, 1.027f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 1f, 1.233f)]
    public void TextSizes(float lineSpacing, float paragraphSpacing, float stretchY, float capHeight, float ascent, float descent, float height, float leading, float paragraphLeading, float stretchX, float characterSpacing, float width)
    {
        _writer.Font = new Font(StandardFonts.Times, 12f) {
            LineSpacing = lineSpacing,
            ParagraphSpacing = paragraphSpacing,
            StretchY = stretchY,
            CharacterSpacing = characterSpacing,
            StretchX = stretchX,
        };
        _writer.TextCapHeight().ShouldBe(capHeight, 0.001f);
        _writer.TextAscent().ShouldBe(ascent, 0.001f);
        _writer.TextDescent().ShouldBe(descent, 0.001f);
        _writer.TextHeight().ShouldBe(height, 0.001f);
        _writer.TextLeading().ShouldBe(leading, 0.001f);
        _writer.TextHeight().ShouldBe(_writer.TextAscent() + _writer.TextDescent() + _writer.TextLeading(), 0.001f);
        _writer.TextParagraphLeading().ShouldBe(paragraphLeading, 0.001f);
        _writer.TextWidth("Hello World!").ShouldBe(width, 0.001f);
    }
    
    [Fact]
    public void Standard()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void CharacterSpacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { CharacterSpacing = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
