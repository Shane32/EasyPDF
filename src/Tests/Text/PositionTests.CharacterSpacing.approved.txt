%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 918/Filter/FlateDecode>>stream
x\x9C\x8D\x97MO\x1B1\x10\x86\xEF\xFB+\xF6\x08\x87\xA4\x1E{\xBF\xDC[\xA9\xDAJ\xA8\x97\xD2\xBDU\x1C\xAA\x10(\x15\x0B\x85F\xEA\xDF\xAFw\xC6\xF3\xE1\x88\x08\x8B\x8FL\xBC3\xE3\xC9\xC3\xF2\xBE\xDE\xE7\xE6bn\xC2\xD0\x8E\xFD\xD0\xCE7\xCD\xA7\xB9\xF9\xD6@\xEB\xD2\xD7\x06\xDA\xD1\xA7o\xD7\xEE\x96\xC6\xB5\x97\xE9\xE7w\xE3\xB6\xD0\xFEk~\\\xA7\xEB7\x8D\xC3\xBC\xAB/9x\xB9{=\xCD{\xB7\xEDS\xB4\xE4\xC8o\xBB\xD8z\x98\xD6\x97n}\x0F\x03\xBE\xEE\x1A\x0Fa\xDBC^\x04M\x05\xAC_\xAF\xAF\xD1&\xAFb\xEE\x86;l\xB8\x05\xF6\xA5e\xDC.\xA7\xE7!v\xCD\xF7\x86\x1B.2\xDA\x03\xAE\xE6&K\xC3\x13\xAD\xCB\xCF\x8Do\xE7\xDDJ\xE9\xDDgh\xC1\xB5\xF3\xAD\x02Z\x13\xC7-L\xED\xBCX\x08nE\xF9\xE3\xEC\xC3\xC3\xFD\xDD\xE3\xB2\x7F<\xBC?\xEF\xDDY\xFBu\x7F{\x98\xCF\xC1\xBB\xB3\xA7?\xE7\xD7\xF3%\xC1\xA6\x11R\x0B\xC1\xE3\x0C\x1F\xEF\xF3\xC7\xC3@\x08\xE5e0\xE9@]\x18\x12\xF4\x98G\xE9\xEBo\xEC\x82\x81@\xA2e\xDA\x9F\xD2y\x16\xA5d\x06K!s\xC2J\xE2\x84\xA3T\x81\xF2\xA9_\xAC&\xF51\xBD\xDB\xBF\x1CsJw*\x8F\x13&{\x1F9\xBE\x91\\q'9\xE1\xC4\xE9@]\x98S\x08\x86S\x80\xDC\x05\x03\xE1D\xCB\xB4iPN\xD8\x859\xD9\xC1\x06\xE1\x84\x95\xF9~r\xB5\x9CR}-\xA4\x8B\x9F\x7F\xF7\x0F\xF7\x8F\xFBcL}'\xD3\xF4\x83\xC1\xD4O\xF9\x03b \x98\xF22\x98t\xA0.\x8C\xA9\x07\x83\xA9\x8B\x0C;ZL\xB4L\x9B\x82b\xC2.\x8C\xC9\x0E\xD6\x09&\xAC$L8J\x15\xA6\xB4G\x0C\xF5\xA4\x9E\x0E\x87\xA7\xE5\x98S\x128\x1Eg\xEC\x0C\xA7q\xC8\x9F\x10\x03\xE1\x94\x97\xC1\xA4\x03uaNC4\x9C\x861w\xC1@8\xD12n\x9A\xD3y\x16\xE5d\x07\xF3\xC2\x09+\x89\x13\x8E\xF2&'Hd}\x12\xF1X\xABQ\xF4_wJ\xA5\xA2jf\xB4*\x1EY\x84c\xA1\xE3Q\x85<\x1A%\x8F*\xE5\xD3hpM}\xEE\x82\x81\xE0\xA2e\xDC4\xA7\xF3,\x8A\xCB\x0E\xA6j\x8E\x95\x84+V\xC99\x84\x84+\xB41\xD4
\x15\xE1z]\xAA\xC0\x19\xE9\x04+\xEA\x00\xAC\xEA\x14\x090\xBE\x00\xB6\x04r/QvWH\xBB\x13mw\xA5\xB8;\xA3\xEE\xCE\xCA\xBB+\xF4\xDDN\xE9\x8C\xC2;\x95x\x1A\xAB\x82^\xD8\xBA\x01\xBB\xD4\xB3;\xA5`\xE0UP\xC1[\xA9\x87\xC0ZO\x91\xD2\x0B\xAA\xF6R\x02\xB9\x97\xD0\xF3V\xF0\xC1\xB3\xE2S\xA4\xF4\xBCj>\x97\xC8\\\x86^1\xA5\xEA>\x95gz\xA1J\xF9!\x19T\x88\xB8Y\xA5\xACe\x80\xAF
\x1Bt*\xB4\xD0Y\x0B\x80\x8E=\x80"\xC5\xD7\xA9\x0BH\x09\xE4^\x82\xAF\x83\xE2\\\x11\xE5OQXA\xBE\x90\xB77f@\xBD\x04_1eg\x8E\x17j\x084\xD6\x9B\xF8"l!-&S\xA8T\xBA\xAB\xFB\xBB_\xA7\x8Fc\x83\xEA/\x0C\xD6\x19``k\xA0H\xF1j\x0ER\x02\xB9\x97\x9E\xCA\xAC?@\xCF\x06A\x91\xE2\xEB\xD5"\xB8D\xE62\xF8\x8A)\xD5&\xA8<\xE3\x1B\xAA\x8CbL\xDCBJ\xEEk\x95\x0F\xF1\x9D\x10\xBE\xC9\x993\xA3\xF5\x09\x98\xD8((Rx\x93Z\x85\x94@\xEE%\xF0F\xEB\x160\xB2]P\xA4\xF0F5\x0C.\x91\xB9\xEC\xC1\xD6N\xA9\xA6A\xE5\x19\xDETe\x1B\xE9\xCC2FlRM\xEE\xA4\xECE\xFB P<\x098y\x14p\xE5\xB3\x80\xB3\x0F\x03\xD64\xA21\x8DX\x98F\x14\xD3\x88\xA5iDc\x1A\xD1\x9AF,L\xC3L\xB9\xC6\xC2.\x9A\xE7\x02We\x1A\xE9\x04\xD4\xB8W\xA5\xEA\x11\xBE#\xD1\xFB\x0FgL\x03\x8C
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001157 00000 n 
0000000015 00000 n 
0000001245 00000 n 
0000001000 00000 n 
0000001310 00000 n 
0000001355 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 